server/fuzz.go

```diff
@@ -0,0 +1,34 @@
+// Copyright 2012-2020 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build gofuzz
+
+package server
+
+func Fuzz(data []byte) int {
```

Comment:
Hey @AdamKorcz, looking at this general structure, Currently you are picking up dummyClient() from parser_test.go. What if I wanted to add dummyRouteClient() as well or wanted to add fuzzing for an altogether different scenario say fuzzing that also includes receiving a message?

Similar to Test, do I just add more functions named `Fuzz...(daba []byte) int` or by switching cases on say the first byte handed to `Fuzz`?
I'm asking as I found one article that suggested as much but couldn't quite find that supported elsewhere.

Reply:
Are you looking to fuzz the parser with the `dummyRouteClient` as well?

Regarding fuzzing and receiving a message: Could you point me to the part of the code that you are thinking of? 

To write more fuzzers, you write new functions named `FuzzSOMETHING (data []byte) int`. 

Might I suggest that in the early stages of fuzzing that I write the fuzzers and merge them in? I will be happy to do it to get things started, so if you have any parts of the code you would like to see fuzzed, please throw it at me and I will look into it.

Regarding running the fuzzers locally: I will add the instructions to do that in comments in the fuzzer, is that okay? - And yes, if you have added `$GOPATH/bin` to your `$PATH` variable it would be `go-fuzz-build && go-fuzz`

Reply:
> Are you looking to fuzz the parser with the dummyRouteClient as well?

Sort of, when looking at the implementation of dummyClient and dummyRouteClient I would add tests where kind has the values ROUTER, GATEWAY, LEAF & CLIENT (client has the value 0), essentially testing if the connection is a client connection, cluster connection, gateway or leaf node connection.
`&client{srv: New(&defaultServerOptions), msubs: -1, mpay: -1, mcl: MAX_CONTROL_LINE_SIZE, kind: CLIENT}`

```
func dummyClient() *client {
	return &client{srv: New(&defaultServerOptions), msubs: -1, mpay: -1, mcl: MAX_CONTROL_LINE_SIZE}
}

func dummyRouteClient() *client {
	return &client{srv: New(&defaultServerOptions), kind: ROUTER}
}
```

> Regarding fuzzing and receiving a message: Could you point me to the part of the code that you are thinking of?
Well, nothing concrete yet, but am wondering how much mileage we'd get out of doing more than just parsing. 
Like parse, but also have a subscriber that is expected to receive one message. 
I want to give a hypothetical example. If the server has no subscriber at all, why look at at a messages subject during fuzzing ever?

Reply:
That said, testing the client connection is the more important one to test.
Code wise I'd happy with what is submitted. 
Maybe rename Fuzz to FuzzClient.